src/parquet/ReadParquet.cpp

```diff
@@ -1,5 +1,407 @@
+
+    // In ReadParquet.cpp, there is a basic look up table for this. But number
+    // of required bytes can be found mathematically:
+    int numbits = ceil(precision*3.321928);  // the magic number is the constant
+                                             // from log2(10^precision)
+    if (numbits%8==0) numbits++;             // add a bit for the sign bit if we
+                                             // are at the byte boundary
+    const auto numbytes = ceil(numbits/8.0);
+
+    while (reader->HasNext() && row_idx < numElems) {
+      PqType value;
+      (void)reader->ReadBatch(1, nullptr, nullptr, &value, &values_read);
+      arrow::Decimal128 v;
+      PARQUET_ASSIGN_OR_THROW(v,
+                              ::arrow::Decimal128::FromBigEndian(value.ptr,
+                                                                 numbytes));
+
+      chpl_ptr[row_idx] = v.ToDouble(0);
+      row_idx+=values_read;
+      num_read+=values_read;
+    }
+
+    return num_read;
+  }
+
+  template<typename Types>
+  int64_t _readShortIntegral() {
+    using ReaderType = typename Types::ReaderType;
+    using ChplType = typename Types::ChplType;
+    using PqType = typename Types::PqType;
+
+    auto chpl_ptr = (ChplType*)chpl_arr;
+    ReaderType* reader = static_cast<ReaderType*>(column_reader.get());
+    *startIdx -= reader->Skip(*startIdx);
+
+    // TODO find better variable names. values_read and num_read are confusing
+    int64_t values_read = 0;
+
+    int64_t num_read = 0;
+    if (not hasNonFloatNulls) {
+      // TODO we can read into the actual Chapel array, and then space the data
+      // out in reverse order. I don't think we need to have this temporary
+      // buffer
+      PqType* tmpArr = (PqType*)malloc(batchSize * sizeof(PqType));
+      while (reader->HasNext() && row_idx < numElems) {
+        if((numElems - row_idx) < batchSize) // adjust batchSize if needed
+          batchSize = numElems - row_idx;
+
+        // Can't read directly into chpl_ptr because it is int64
+        reader->ReadBatch(batchSize, nullptr, nullptr, tmpArr, &values_read);
+
+        for (int64_t j = 0; j < values_read; j++) {
+          chpl_ptr[row_idx+j] = (ChplType)tmpArr[j];
+        }
+
+        row_idx+=values_read;
+        num_read+=values_read;
+      }
+      free(tmpArr);
+    }
+    else {
+      PqType tmp;
+      // Engin: we don't seem to use this anywhere, but passing nullptr instead
+      // of this causes errors. Why?
+      int16_t definition_level; // nullable type and only reading single
+                                // records in batch
+      while (reader->HasNext() && row_idx < numElems) {
+        reader->ReadBatch(1, &definition_level, nullptr, &tmp, &values_read);
+        // if values_read is 0, that means that it was a null value
+        if(values_read == 0) {
+          where_null_chpl[row_idx] = true;
+        }
+        else {
+          chpl_ptr[row_idx] = (ChplType)tmp;
+        }
```

Comment:
prefer ternary

Reply:
This is actually the snippet I was referring to above as something that's not very suitable for ternary. Note that the two branches assign to different things. I can make use of ternary operator here, but that would feel like an abuse of it. As, I think this should be kept as is, I prefer to keep other similar post-processing related conditionals the same for the time being.